### PR TITLE
Olap props writeback

### PR DESCRIPTION
**Pull request by zhoney**

### CODE REVIEW

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java [hunk 1]
```diff
@@ -74,6 +75,15 @@ public Vertex create(Vertex vertex) {
         return result.readList(this.type(), Vertex.class);
     }
 
+    public int update(BatchOlapPropertyRequest request) {
+        this.client.checkApiVersion("0.56", "batch property update");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:79]: olap property batch update

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java [hunk 2]
```diff
@@ -74,6 +75,15 @@ public Vertex create(Vertex vertex) {
...
+        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
+        headers.putSingle("Content-Encoding", BATCH_ENCODING);
+        String path = String.join("/", this.path(), "olap/batch");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:82]: is there olap property single update

File: src/main/java/com/baidu/hugegraph/structure/graph/BatchOlapPropertyRequest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        return String.format("BatchOlapPropertyRequest{vertices=%s," +
+                             "propertyKey=%s}",
+                             this.vertices, this.propertyKey);
+    }
+
+    public static class Builder {
+
+        private BatchOlapPropertyRequest req;
+
+        public Builder() {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/structure/graph/BatchOlapPropertyRequest.java:54]: set private

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java (hunk 2 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:82]: is there olap property single update?

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java (hunk 1 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:79]: prefer "olap property batch update"

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java (hunk 2 above)
Comment (zhoney) [reply src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:82]: there is no single update in server api so far. need to add one?

**Comment by zhoney** [comment]
also, server need to ignore olap field in PropertyKeyAPI or add olap field(but not use)

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java [hunk 4]
```diff
@@ -74,6 +75,15 @@ public Vertex create(Vertex vertex) {
         return result.readList(this.type(), Vertex.class);
     }
 
+    public int update(BatchOlapPropertyRequest request) {
+        this.client.checkApiVersion("0.56", "olap property batch update");
+        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
+        headers.putSingle("Content-Encoding", BATCH_ENCODING);
+        String path = String.join("/", this.path(), "olap/batch");
+        RestResult result = this.client.put(path, null, request, headers);
+        return (int) result.readObject(Map.class).get("size");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:84]: check response: size instanceof Int

**Review by javeme — CHANGES_REQUESTED** [review]

**Review by javeme — APPROVED** [review]

File: src/main/java/com/baidu/hugegraph/api/graphs/GraphsAPI.java [hunk 5]
```diff
@@ -77,6 +78,32 @@ public GraphMode mode(String graph) {
...
+
+    public GraphReadMode readMode(String graph) {
+        String path = String.join("/", this.path(), graph);
+        RestResult result =  this.client.get(path, "graph_read_mode");
+        @SuppressWarnings("unchecked")
+        Map<String, String> readMode = result.readObject(Map.class);
+        String value = readMode.get("graph_read_mode");
+        if (value == null) {
+            throw new InvalidResponseException(
+                      "graph_read_mode", "graph_read_mode response");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/api/graphs/GraphsAPI.java:97]: improve message

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java [hunk 6]
```diff
@@ -74,6 +76,18 @@ public Vertex create(Vertex vertex) {
...
     }
 
+    public int update(BatchOlapPropertyRequest request) {
+        this.client.checkApiVersion("0.56", "olap property batch update");
+        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
+        headers.putSingle("Content-Encoding", BATCH_ENCODING);
+        String path = String.join("/", this.path(), "olap/batch");
+        RestResult result = this.client.put(path, null, request, headers);
+        Object size = result.readObject(Map.class).get("size");
+        E.checkState(size instanceof Integer,
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:86]: also throw InvalidResponseException, or add method invalidResponse(message, args)

File: src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java [hunk 7]
```diff
@@ -74,6 +76,18 @@ public Vertex create(Vertex vertex) {
...
+                     "The size must be int, but got: %s", size);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/api/graph/VertexAPI.java:87]: The size in response...

File: src/main/java/com/baidu/hugegraph/api/graphs/GraphsAPI.java [hunk 8]
```diff
@@ -77,6 +78,32 @@ public GraphMode mode(String graph) {
         }
     }
 
+    public void readMode(String graph, GraphReadMode readMode) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/api/graphs/GraphsAPI.java:81]: check version first

**Review by javeme — CHANGES_REQUESTED** [review]

**Review by javeme — APPROVED** [review]

**Review by javeme — APPROVED** [review]

**Review by Linary — APPROVED** [review]

File: src/main/java/com/baidu/hugegraph/structure/schema/PropertyKey.java [hunk 9]
```diff
@@ -37,13 +37,16 @@
     private Cardinality cardinality;
     @JsonProperty("aggregate_type")
     private AggregateType aggregateType;
+    @JsonProperty("olap")
+    private Boolean olap;
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/structure/schema/PropertyKey.java]: change to readfrequecy?

**Pull request merged**